core/payments/ondemand/on_demand_ledger.go

```diff
@@ -0,0 +1,167 @@
+// - If this is not respected, the entity doing the accounting will diverge from the correct ledger state
+//
+// TODO(litt3): There are some improvements that may be made to the on-demand logic in the future, to permit a larger
+// number of concurrent on-demand dispersals.
+type OnDemandLedger struct {
+	// total deposits available for the account in wei
+	totalDeposits *big.Int
+	// price per symbol in wei
+	pricePerSymbol *big.Int
+	// minimum number of symbols to bill
+	minNumSymbols uint64
+	// stores the cumulative payment for this ledger in wei
+	cumulativePaymentStore CumulativePaymentStore
+}
+
+// Constructs a new OnDemandLedger, backed by the input CumulativePaymentStore
+func NewOnDemandLedger(
+	totalDeposits *big.Int,
+	pricePerSymbol *big.Int,
+	minNumSymbols uint64,
+	cumulativePaymentStore CumulativePaymentStore,
+) (*OnDemandLedger, error) {
+	if totalDeposits == nil {
+		return nil, errors.New("totalDeposits cannot be nil")
+	}
+	if totalDeposits.Sign() < 0 {
+		return nil, errors.New("totalDeposits cannot be negative")
+	}
+
+	if pricePerSymbol == nil {
+		return nil, errors.New("pricePerSymbol cannot be nil")
+	}
+	if pricePerSymbol.Sign() < 0 {
+		return nil, errors.New("pricePerSymbol cannot be negative")
+	}
+
+	if cumulativePaymentStore == nil {
+		return nil, errors.New("cumulativePaymentStore cannot be nil")
+	}
+
+	return &OnDemandLedger{
+		totalDeposits:          totalDeposits,
+		pricePerSymbol:         pricePerSymbol,
+		minNumSymbols:          minNumSymbols,
+		cumulativePaymentStore: cumulativePaymentStore,
+	}, nil
+}
+
+// Debit the on-demand account with the cost of a dispersal, based on the number of symbols.
+//
+// Returns (cumulativePayment, nil) if the account has sufficient funds to perform the debit.
+// The returned cumulativePayment represents the new total amount spent from this account, including this blob.
+//
+// Returns (nil, error) if an error occurs. Possible errors include:
+//   - [QuorumNotSupportedError]: requested quorums are not supported for on-demand payments
+//   - [InsufficientFundsError]: the debit would exceed the total deposits available
+//   - Generic errors for all other unexpected behavior
+//
+// If the account doesn't have sufficient funds to accommodate the debit, the cumulative payment
+// IS NOT updated, i.e. a failed debit doesn't modify the payment state.
+func (odl *OnDemandLedger) Debit(
+	ctx context.Context,
+	symbolCount uint32,
+	quorums []core.QuorumID,
+) (*big.Int, error) {
+	if symbolCount == 0 {
+		return nil, errors.New("symbolCount must be > 0")
+	}
+
+	err := checkForOnDemandSupport(quorums)
+	if err != nil {
+		return nil, err
+	}
+
+	blobCost := odl.computeCost(symbolCount)
+
+	newCumulativePayment, err := odl.cumulativePaymentStore.AddCumulativePayment(ctx, blobCost, odl.totalDeposits)
+	if err != nil {
+		return nil, fmt.Errorf("add cumulative payment: %w", err)
+	}
```

Comment:
should we panic here if a negativeBalanceError is returned? https://github.com/Layr-Labs/eigenda/pull/1881/files#diff-4e474963d9f8251df963bca81d3bc27580eec9e636c136552cb926afccc26bcbR54

After discussing with Cody, I agree that we shouldn't panic inside public methods because we don't know who will be using them. But on the outside, where its called from, if that error is returned, there's literally a bug right? We're trying to debit a negative value which is a bug? And at this point the software is literally broken and god knows that the accounting is doing? So we should panic iiuc?

Reply:
Btw this is a follow-up to @cody-littley's answer here: https://github.com/Layr-Labs/eigenda/pull/1881#discussion_r2282803025